pkg/workloads/containerd/watcher.go

```diff
@@ -53,106 +52,80 @@ const (
 	eventQueueBufferSize = 100
 )
 
-// containerEvents holds per-container queues for events
-type containerEvents struct {
-	lock.Mutex
-	events map[string]chan dTypesEvents.Message
-}
-
-func (ce *containerEvents) enqueueByContainerID(e dTypesEvents.Message) {
-	ce.Lock()
-	defer ce.Unlock()
-
-	if _, found := ce.events[e.Actor.ID]; !found {
-		q := make(chan dTypesEvents.Message, eventQueueBufferSize)
-		ce.events[e.Actor.ID] = q
-		go processContainerEvents(e.Actor.ID, q)
-	}
-	ce.events[e.Actor.ID] <- e
-}
-
-func (ce *containerEvents) reapEmpty() {
-	ce.Lock()
-	defer ce.Unlock()
-
-	for id, q := range ce.events {
-		if len(q) == 0 {
-			close(q)
-			delete(ce.events, id)
-		}
-	}
-}
-
 func shortContainerID(id string) string {
 	return id[:10]
 }
 
 // EnableEventListener watches for docker events. Performs the plumbing for the
 // containers started or dead.
 func EnableEventListener() error {
-	eventQueue := containerEvents{events: make(map[string]chan dTypesEvents.Message)}
-	since := time.Now()
-	syncWithRuntime()
+	ws := newWatcherState(eventQueueBufferSize)
+	syncWithRuntime(ws)
 
+	since := time.Now()
 	eo := dTypes.EventsOptions{Since: strconv.FormatInt(since.Unix(), 10)}
 	r, err := dockerClient.Events(ctx.Background(), eo)
 	if err != nil {
 		return err
 	}
 
-	go listenForDockerEvents(&eventQueue, r)
+	go listenForDockerEvents(ws, r)
 
 	// start a go routine which periodically synchronizes containers
 	// managed by the local container runtime and checks if any of them
 	// need to be managed by Cilium. This is a fall back mechanism in case
 	// an event notification has been lost.
-	go func() {
+	go func(state *watcherState) {
 		for {
 			time.Sleep(syncRateDocker)
-			syncWithRuntime()
 
-			log.Debug("Reaping empty event queues")
-			eventQueue.reapEmpty()
+			log.Debug("Reaping empty containerd watcher event queues")
+			state.reapEmpty()
+
+			log.Debug("Periodic watcher synchronization to containerd runtime")
```

Comment:
Having this message every 30 seconds can be too verbose, do we really need it?

Reply:
It was printing that before. I would prefer to only print a message when we find an unwatched container, which I added in `syncWithRuntime`. I'm happy to remove both this one and the reaping one above since neither seems very useful.

Reply:
OK, removed these debug lines after discussing it in person.